nuxeo-core/nuxeo-core-storage-dbs/src/main/java/org/nuxeo/ecm/core/storage/dbs/DBSTransactionState.java

```diff
@@ -402,31 +422,156 @@ public void updateAncestors(String id, int ndel, Object[] ancestorIds) {
         }
     }
 
+    protected int getReadAclsAsyncThreshold() {
+        boolean enabled = Boolean.parseBoolean(
+                Framework.getProperty(READ_ACL_ASYNC_ENABLED_PROPERTY, READ_ACL_ASYNC_ENABLED_DEFAULT));
+        if (enabled) {
+            return Integer.parseInt(
+                    Framework.getProperty(READ_ACL_ASYNC_THRESHOLD_PROPERTY, READ_ACL_ASYNC_THRESHOLD_DEFAULT));
+        } else {
+            return 0;
+        }
+    }
+
     /**
      * Updates the Read ACLs recursively on a document.
      */
     public void updateTreeReadAcls(String id) {
         // versions too XXX TODO
-        Set<String> ids = getSubTree(id, null, null);
-        ids.add(id);
-        ids.forEach(this::updateDocumentReadAcls);
+
+        save(); // flush everything to the database
+
+        // update the doc itself
+        updateDocumentReadAcls(id);
+
+        // check if we have a small enough number of descendants that we can process them synchronously
+        int limit = getReadAclsAsyncThreshold();
+        Set<String> ids = getSubTree(id, null, null, limit);
+        if (limit == 0 || ids.size() < limit) {
+            // update all descendants synchronously
+            ids.forEach(this::updateDocumentReadAcls);
+        } else {
+            // update the direct children synchronously, the rest asynchronously
+
+            // update the direct children (with a limit in case it's too big)
+            String nxql = String.format("SELECT ecm:uuid FROM Document WHERE ecm:parentId = '%s'", id);
+            Principal principal = new SystemPrincipal(null);
+            QueryFilter queryFilter = new QueryFilter(principal, null, null, null, Collections.emptyList(), limit, 0);
```

Comment:
instead of duplicating the logic, it is possible to run a FindReadAclsWork synchronously

Reply:
This one only checks the direct children and collects the ids, but does not launch new works. There's really not much duplication either.